### ISSUE TITLE
cohttp-eio hangs

### ISSUE THREAD

**Issue by craff** [mirage/ocaml-cohttp#953]
The simplest example hangs if I do more requests than the number of domains. The bug is against latest opam version with ocaml 5.0.0-beta2.

Step to reproduce:
dune build
dune exec ./bug.exe
./bench.sh http://localhost:8080/ N 0

N should be 1+the number of core (recommended_domain_count)

Attaching a file keep failing on gitbug (sorry github ;-). I put the file inline 

--------------------------  bug.ml -----------------
module S = Cohttp_eio.Server

let () =
  let handler (_request,_buf,_addr) =
    let data = String.make (1024*1024) 'X' in
    S.text_response data
  in
  let port = 8080 in
  let domains = Domain.recommended_domain_count () in
  Eio_main.run @@ fun env -> S.run ~domains ~port env handler
------------------------------ dune -------------------------
(executable
 (name bug)
 (modules :standard)
 (libraries eio eio_main cohttp-eio))
------------------------ bench.sh --------------------------
#!/bin/bash

url=$1
nb=$2
sleep_time=$3

for (( c=1; c<=$nb; c++ )); do
  f=$(mktemp)
  (curl -s $url > $f; stat -c %s $f; rm $f) &
  sleep $sleep_time
done

wait $(jobs -p)
--------------------------------------------------------------------

**Comment by mseri** [comment]
Thanks for the report and reproduction mteril. Ping @bikallem

**Comment by bikallem** [comment]
Ack.

**Comment by bikallem** [comment]
@craff I am unable to reproduce the error (the hang). 

```
$ ./bench.sh localhost:8080 3 0
/tmp/nix-shell.0dwbKW/tmp.y7Ek4nEStT
1048576
/tmp/nix-shell.0dwbKW/tmp.prkvJwpBu4
1048576
/tmp/nix-shell.0dwbKW/tmp.PCpqEqLPzW
1048576
```

**Comment by craff** [comment]
Maybe it is only a bug on the current opam version. There is an issue #883 that was "hanging with curl" that may not be pushed to opam yet ?

The problem is that opam pin add . failed for me on the current github version.

**Comment by talex5** [comment]
I see this with `EIO_BACKEND=luv`. I suspect this is because we use a separate Luv event loop for each domain, but cohttp is sharing the socket between domains. Sockets get attached to an event loop when they are created, not when you wait on them. Possibly we should run a single Luv event loop in a sys-thread and dispatch events to domains from that.

**Comment by talex5** [comment]
> Ideally each domain would have its own listener socket, bound to the same address and port using SO_REUSEADDR and SO_REUSEPORT

libuv doesn't support `SO_REUSEPORT` however :-(

Looking a bit closer, I think this is what's happening:

1. We get notified of a new connection: https://github.com/ocaml-multicore/eio/blob/50abf52dd3fab1565114801d9edd8d7e13dabab0/lib_eio_luv/eio_luv.ml#L796-L800
2. We signal that a new FD is ready and return from the accept handler.
3. libuv notices that we didn't take the FD yet and decides we don't want any more.
   It disables watching the listening socket: https://github.com/libuv/libuv/blob/988f2bfc4defb9a85a536a3e645834c161143ee0/src/unix/stream.c#L569-L576
4. One of our worker domains accepts the FD, but that's the last one we'll get.

Libuv should resume watching the listening socket when we accept, but I guess it gets confused because the accept is done from a different domain.

**Comment by bikallem** [comment]
> Ideally each domain would have its own listener socket, bound to the same address and port using SO_REUSEADDR and SO_REUSEPORT, this way they can all call accept on their own descriptors and the kernel should round-robin the connections. Decades ago support for this was fiddly,

Slightly off topic: I experimented with this approach thinking it would enhance performance (i.e. we would get free load balancing from linux kernel among many of the CPU cores). The benchmarks however didn't reveal any benefits whatsoever. 

I think using SO_REUSEPORT is only usable/beneficial if you have multiple OS processes. It doesn't seem to matter much if you use it in a single process setting.

**Comment by mseri** [comment]
Should we just remove it then?

**Comment by bikallem** [comment]
> Should we just remove it then?

Yes. The `Server.run` is due some improvements as discussed in https://github.com/mirage/ocaml-cohttp/issues/961 and this includes removing `SO_REUSEPORT` usage.

**Comment by haesbaert** [comment]
That works, but there are many things involved, multiple sockets with
reuseaddr/port work in a push manner. The kernel round robins the n XT
connection, which means that your next connection might wait if that domain
is busy. If it's selected or epoll/level triggered, it also causes
thundering storm, as in the "file description" not descriptor is considered
readable and everyone wakes up. Scaling all of this requires thinking about
all this little cases and how they play a part.
